### PR TITLE
ddl: fix add unique index on partitioned table (by RANGE COLUMNS) failed

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <heng@lonng.org>

### What problem does this PR solve? 

```
create table t(
    a int,
    b varchar(100),
    c int,
    INDEX idx_c(c))
PARTITION BY RANGE COLUMNS( a ) (
    PARTITION p0 VALUES LESS THAN (6),
    PARTITION p1 VALUES LESS THAN (11),
    PARTITION p2 VALUES LESS THAN (16),
    PARTITION p3 VALUES LESS THAN (21)
);
```
```
mysql> alter table t add unique index idx_a(a);
ERROR 1105 (HY000): line 1 column 7 near ""
```

### What is changed and how it works?

The `Expr` field in `PartitionInfo` is not always be assigned a value (e.g: `PARTITION BY RANGE COLUMNS(...)`).
This PR use columns in `PartitionInfo` directly if the `PartitionInfo.Expr` is an empty string.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Fix add unique index on the partitioned table (by RANGE COLUMNS(...)) failed.

### CODE REVIEW

File: ddl/partition.go [hunk 1]
```diff
@@ -620,11 +620,28 @@ func checkRangePartitioningKeysConstraints(sctx sessionctx.Context, s *ast.Creat
...
+	if partExpr := pi.Expr; partExpr != "" {
+		// Parse partitioning key, extract the column names in the partitioning key to slice.
+		partCols, err = extractPartitionColumns(partExpr, tblInfo)
+		if err != nil {
+			return err
+		}
+	} else {
+		partCols = make([]*model.ColumnInfo, 0, len(pi.Columns))
+		for _, col := range pi.Columns {
+			for _, info := range tblInfo.Columns {
```
Comment (zimulala) [inline ddl/partition.go]: Could we use `getColumnInfoByName`？

File: ddl/db_integration_test.go [hunk 2]
```diff
@@ -1647,6 +1647,28 @@ func (s *testIntegrationSuite3) TestAlterAlgorithm(c *C) {
...
+	INDEX idx_c(c))
+	PARTITION BY RANGE COLUMNS( a ) (
+		PARTITION p0 VALUES LESS THAN (6),
+		PARTITION p1 VALUES LESS THAN (11),
+		PARTITION p2 VALUES LESS THAN (16),
+		PARTITION p3 VALUES LESS THAN (21)
+	)`)
+	s.tk.MustExec("alter table t add unique index idx_a(a)")
+	s.tk.MustExec("alter table t add unique index idx_ac(a, c)")
+	assertErrorCode(c, s.tk, "alter table t add unique index idx_b(b)", mysql.ErrUniqueKeyNeedAllFieldsInPf)
```
Comment (tiancaiamao) [inline ddl/db_integration_test.go]: Would you like to add those code:

```
s.tk.MustExec("insert into t values (4, "xxx", 4)")
s.tk.MustExec("insert into t values (4, "xxx", 9)")   // Note the repeated 4
s.tk.MustExec("insert into t values (17, "xxx", 12)")
err := Exec("alter table t add unique index idx_a(a)")  // And see the result
```

Comment (lonng) [reply ddl/db_integration_test.go]: tests added

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/partition.go [hunk 3]
```diff
@@ -620,11 +621,28 @@ func checkRangePartitioningKeysConstraints(sctx sessionctx.Context, s *ast.Creat
...
 }
 
-func checkPartitionKeysConstraint(sctx sessionctx.Context, partExpr string, idxColNames []*ast.IndexColName, tblInfo *model.TableInfo) error {
-	// Parse partitioning key, extract the column names in the partitioning key to slice.
-	partCols, err := extractPartitionColumns(partExpr, tblInfo)
-	if err != nil {
-		return err
+func checkPartitionKeysConstraint(pi *model.PartitionInfo, idxColNames []*ast.IndexColName, tblInfo *model.TableInfo) error {
+	var (
+		partCols []*model.ColumnInfo
```
Comment (crazycs520) [inline ddl/partition.go]: `partCols ` is useless? can we remove it?

**Comment by lonng** [comment]
Cherry-pick to release-3.0 after https://github.com/pingcap/tidb/pull/11954 merged

**Pull request merged**